### DISCUSSION TITLE
The trainers runs a single validation step after resume (not sanity)

### DISCUSSION THREAD

**Discussion by cdancette** [Lightning-AI/pytorch-lightning#18110] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
I am encountering a weird behavior using lightning (I use the lightningCLI as well).

When I resume a training that has failed, with `python train.py fit`, the trainer runs a single validation step that is not a sanity check (I disabled sanity checks, and also the state of the trainer is `RunningStage.VALIDATING`).

I run my main training with val_check_interval=0.5.

This is a problem, as it saves in my logs a metric point for this single batch as if it was an evaluation on the full validation set. 
I have metrics curve like this: when I resume, there is a wrong point (much higher here than the other ones). 

<img width="376" alt="image" src="https://github.com/Lightning-AI/lightning/assets/10550327/472c2c71-3d5b-43d6-9bc3-1ada30743727">

Do you know what could cause this issue and how to get around it ?

**Comment by arnaudstiegler** [comment]
Seeing a similar behavior with version `2.0.2` and `2.0.7` (latest). Might be worth opening an issue

**Reply by jojonki** [reply]
I also faced this problem at `2.2.0`. I set `num_sanity_val_steps=0` at Trainer and use an epoch-end checkpoint.

**Comment by jojonki** [comment]
I debugged the phenomenon step by step. The issue is that during the restarting status, 0 steps of training and 1 step of validation are performed. To fix this, it is necessary to correct the restart logic, which seems to require a deep understanding of the PL code (I gave up and decided to allow this 1-step validation).

When loading a checkpoint, `restarting` is set to True.
https://github.com/Lightning-AI/pytorch-lightning/blob/1439da41b25a1efac581db35433e19e9d2b736d2/src/lightning/pytorch/loops/loop.py#L84

When training begins, it first checks if `restarting is True` and if validation is defined. If resuming training, it matches this condition and skips training. The Training Epoch progress bar ends at 0%.
https://github.com/Lightning-AI/pytorch-lightning/blob/1439da41b25a1efac581db35433e19e9d2b736d2/src/lightning/pytorch/loops/training_epoch_loop.py#L199-L201

Next, it enters the validation loop.
https://github.com/Lightning-AI/pytorch-lightning/blob/1439da41b25a1efac581db35433e19e9d2b736d2/src/lightning/pytorch/loops/fit_loop.py#L197-L200

In EvaluationLoop's `setup_data`, it returns immediately, and `data_fetcher` remains None.
https://github.com/Lightning-AI/pytorch-lightning/blob/1439da41b25a1efac581db35433e19e9d2b736d2/src/lightning/pytorch/loops/evaluation_loop.py#L147-L148

Following the `reset`, data is "prefetched only once".
https://github.com/Lightning-AI/pytorch-lightning/blob/1439da41b25a1efac581db35433e19e9d2b736d2/src/lightning/pytorch/loops/evaluation_loop.py#L219
https://github.com/Lightning-AI/pytorch-lightning/blob/1439da41b25a1efac581db35433e19e9d2b736d2/src/lightning/pytorch/loops/fetchers.py#L95

In EvaluationLoop, `self._restarting` becomes `False`, and normal training begins thereafter.
https://github.com/Lightning-AI/pytorch-lightning/blob/1439da41b25a1efac581db35433e19e9d2b736d2/src/lightning/pytorch/loops/fit_loop.py#L210

**Comment by Youyoun** [comment]
Hello (:wave: @cdancette), I am encountering the same issue. A single validation step is performed after a mid-epoch checkpoint is loaded, where the stage is not set as sanity according to Trainer. Any news or issue related to this discussion ?

**Reply by cdancette** [reply]
haha just noticed this, hey @Youyoun ! I don't know if it's fixed yet, I haven't been in the same training situation in a few months !

**Comment by korotaS** [comment]
Seems that this issue was already present a long ago (#11504) and it was even fixed (#11552), but if we look on current code is seems that the fix is gone... 

The fix in version 1.5.10 (lines 534-535):
https://github.com/Lightning-AI/pytorch-lightning/blob/9ebdc52ec631df92ca7cfc1ba852801dd36d3864/pytorch_lightning/loops/epoch/training_epoch_loop.py#L530-L538

That same code in the next version 1.6.0 ant later versions doesn't have the fix: 
https://github.com/Lightning-AI/pytorch-lightning/blob/44e3edbf7a3fffd3069e2bc24c6ee143de79944b/pytorch_lightning/loops/epoch/training_epoch_loop.py#L522-L528

**Comment by theodorblackbird** [comment]
Facing the same issue, it is particularly annoying when using checkpoint callback with save_top_k since it tracks possibly wrong "best metric".

**Comment by shirondru** [comment]
I'm also facing this issue

**Comment by SeminKim** [comment]
Also facing the same issue. I think fixing this correctly is a bit tricky, so I just ended up with dirty workaround:
manually exit train epoch loop with `on_validation_model_zero_grad` hook, which is called just before `val_loop.run()` in `train_loop.on_advance_end()`. Worked for me with Lightning 2.1.3.
```
    def on_validation_model_zero_grad(self):
        '''
        Small hack to avoid first validation on resume. 
        This will NOT work if the gradient accumulation step should be performed at this point.
        '''
        super().on_validation_model_zero_grad()
        if self.trainer.ckpt_path is not None and getattr(self, '_restarting_skip_val_flag', True):
            self._restarting_skip_val_flag = False
            raise StopIteration
```

**Reply by ethanhe42** [reply]
this works for me but somehow messed up global_step counter

**Reply by ethanhe42** [reply]
here's a better fix without StopIteration 
```
    def on_validation_model_zero_grad(self) -> None:
        '''
        Small hack to avoid first validation on resume. 
        This will NOT work if the gradient accumulation step should be performed at this point.
        '''
        super().on_validation_model_zero_grad()
        if self.trainer.ckpt_path is not None and getattr(self, '_restarting_skip_val_flag', True):
            self.trainer.sanity_checking = True
            self._restarting_skip_val_flag = False
```

**Comment by magehrig** [comment]
I also encountered this problem. Is there an github issue open for this? This definitely deserves attention

**Comment by ethanhe42** [comment]
this is very annoying. deserves more attention

**Reply by Youyoun** [reply]
I created an issue but wasn't addressed yet: https://github.com/Lightning-AI/pytorch-lightning/issues/20288

**Reply by roxas6path** [reply]
This has also been raised in #20277
